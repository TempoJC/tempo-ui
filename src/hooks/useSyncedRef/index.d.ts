import type { Ref, MutableRefObject } from "react";

declare function useSyncedRef<T>(ref: Ref<T>): MutableRefObject<T>;

export { useSyncedRef as default };
