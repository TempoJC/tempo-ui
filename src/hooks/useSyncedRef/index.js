export { default } from "./useSyncedRef.js";
