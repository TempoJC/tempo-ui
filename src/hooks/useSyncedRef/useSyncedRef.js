import { useRef } from "react";

const EMPTY_REF = Symbol("EMPTY_REF");

const useSyncedRef = (ref) => {
  // Inner ref to use and sync with outside ref
  // If the outside ref is already set, sync currents
  // Caveat: if ref is a function, there won't be
  // synced but... it doesn't make much sense to
  // have a useSyncedRef in a component if you are
  // not going to set the ref yourself.
  const innerRef = useRef(ref && ref.current);

  // Ref to memoize the "fake ref API" to ensure that
  // like any real ref we always use the same instance
  const memoRef = useRef();

  // Ref to check if the ref passed to useSyncedRef changed
  // and the memoized ref needs to change as well.
  // Initialized to an special symbol.
  const prevRef = useRef(EMPTY_REF);

  // The  memoRef is always initialized because we
  // check against the symbol. Checks againts null
  // or undefined wouldn't suffice.
  if (prevRef.current !== ref) {
    // Cache the new previousRef.
    prevRef.current = ref;

    // Create the fake Ref. A ref if you look at the
    // react internals is just an object with a current
    // property.
    memoRef.current = {
      // The getter always return the innerRef. The synced
      // ref works either we receive a ref or not.
      get current() {
        return innerRef.current;
      },

      // The setter
      set current(value) {
        // Sets the inner ref to the new current
        innerRef.current = value;

        // If no external ref given, that's all.
        if (!ref) {
          return;
        }

        // If the external ref is a function. Invoke
        // the function with the new value
        if (typeof ref === "function") {
          ref(value);
          return;
        }

        // If the external ref is an object, set its
        // current
        ref.current = value;
      },
    };
  }

  // Always return the memoized ref
  return memoRef.current;
};

export default useSyncedRef;
