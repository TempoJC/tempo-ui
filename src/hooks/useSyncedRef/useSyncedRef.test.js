import React, { createRef } from "react";
import useSyncedRef from "./useSyncedRef.js";
import { render } from "@testing-library/react";

describe("useSyncedRef", () => {
  it("works with object refs", () => {
    let syncedRef;
    let ref = createRef();

    const Comp = () => {
      syncedRef = useSyncedRef(ref);

      return <div ref={syncedRef}>the node</div>;
    };

    render(<Comp />);

    expect(ref.current).toMatchSnapshot();
    expect(syncedRef.current).toBe(ref.current);
  });

  it("works with function refs", () => {
    let syncedRef;
    let node;
    const ref = (r) => {
      node = r;
    };

    const Comp = () => {
      syncedRef = useSyncedRef(ref);

      return <div ref={syncedRef}>the node</div>;
    };

    render(<Comp />);

    expect(node).toMatchSnapshot();
    expect(syncedRef.current).toBe(node);
  });
});
