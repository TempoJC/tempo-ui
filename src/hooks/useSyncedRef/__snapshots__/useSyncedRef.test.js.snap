// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`useSyncedRef works with function refs 1`] = `
<div>
  the node
</div>
`;

exports[`useSyncedRef works with object refs 1`] = `
<div>
  the node
</div>
`;
